### PR TITLE
Improve partial inference of key container operations

### PR DESCRIPTION
**Pull request by timholy**
`resize!` and `sizehint!` are big targets for invalidation, and quite a lot of code was calling these methods with poor inference about the size. Since the size is just `Int`, asserting that eliminates numerous vulnerabilities. Likewise, bounds-checking always works with `Bool`. Finally, add a short-circuit for `eltype(::Tuple{T})`.

### CODE REVIEW

File: base/abstractset.jl [hunk 1]
```diff
@@ -86,7 +86,7 @@ max_values(::Type{Bool}) = 2
 max_values(::Type{Nothing}) = 1
 
 function union!(s::AbstractSet{T}, itr) where T
-    haslength(itr) && sizehint!(s, length(s) + length(itr))
+    haslength(itr) && sizehint!(s, length(s) + length(itr)::Int)
```
Comment (JeffBezanson) [inline base/abstractset.jl:89]: Probably not safe.

File: base/iddict.jl [hunk 2]
```diff
@@ -162,6 +162,7 @@ function get!(default::Callable, d::IdDict{K,V}, @nospecialize(key)) where {K, V
         if !isa(val, V)
             val = convert(V, val)
         end
+        val::V
```
Comment (JeffBezanson) [inline base/iddict.jl:165]: Does it still work if you put this on the `convert(V, val)` call instead? That seems better.

File: base/dict.jl [hunk 3]
```diff
@@ -108,7 +108,7 @@ end
 Dict{K,V}(p::Pair) where {K,V} = setindex!(Dict{K,V}(), p.second, p.first)
 function Dict{K,V}(ps::Pair...) where V where K
     h = Dict{K,V}()
-    sizehint!(h, length(ps))
+    sizehint!(h, length(ps)::Int)
```
Comment (JeffBezanson) [inline base/dict.jl:111]: This is surprising --- can we really not infer this?

File: base/array.jl [hunk 4]
```diff
@@ -984,7 +984,7 @@ push!(a::AbstractVector, iter...) = append!(a, iter)
 function _append!(a, ::Union{HasLength,HasShape}, iter)
     n = length(a)
     i = lastindex(a)
-    resize!(a, n+length(iter))
+    resize!(a, n+length(iter)::Int)
```
Comment (JeffBezanson) [inline base/array.jl:987]: Also this and the next one; arbitrary iterators might not have `Int` length.

File: base/dict.jl (hunk 3 above)
Comment (timholy) [reply base/dict.jl:111]: I can't recall where this came up. I might have just been routinely annotating all `sizehint!` calls. I can't find any evidence that we need this.

File: base/array.jl (hunk 4 above)
Comment (timholy) [reply base/array.jl:987]: You don't mind converting to Int and then asserting?

**Comment by timholy** [comment]
I threw in a random `show_unquoted` fix that I failed to notice earlier.

**Review by KristofferC — APPROVED** [review]
LGTM (I didn't understand the `base/tuple.jl` part though). Could maybe use comments there?

**Pull request merged**